java-client/src/main/java/org/opensearch/client/opensearch/generic/GenericBody.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.client.opensearch.generic;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import javax.annotation.Nullable;
+
+/**
+ * Generic HTTP request / response body. It is responsibility of the caller to close the body instance
+ * explicitly (or through {@link GenericResponse} instance) to release all associated streams.
+ */
+public interface GenericBody extends AutoCloseable {
+    final int DEFAULT_BUFFER_SIZE = 8192;
+
+    /**
+     * Constructs the generic response body out of {@link InputStream} with assumed content type
+     * @param body response body stream
+     * @param contentType content type
+     * @return generic response body instance
+     */
+    static @Nullable GenericBody from(@Nullable final InputStream body, @Nullable final String contentType) {
+        if (body == null) {
+            return null;
+        } else {
+            return new GenericInputStreamBody(body, contentType);
+        }
+    }
+
+    /**
+     * Constructs the generic response body out of {@link InputStream} with assumed content type
+     * @param body response body stream
+     * @param contentType content type
+     * @return generic response body instance
+     */
+    static @Nullable GenericBody from(@Nullable final byte[] body, @Nullable final String contentType) {
+        if (body == null) {
+            return null;
+        } else {
+            return new GenericByteArrayBody(body, contentType);
+        }
+    }
+
+    /**
+     * Content type of this body
+     * @return content type
+     */
+    String contentType();
+
+    /**
+     * Gets the body as {@link InputStream}
+     * @return body as {@link InputStream}
+     */
+    InputStream body();
```

Comment:
@dblock one decision which is doubt at the moment is using `InputStream` vs `byte[]` to deal with response / request payload. Usage of `InputStream`  has a lot of benefits for a price of requiring to be more careful to deal with (the responses should be closed upon consumption). 

The `byte[]` would make iAPI much simpler but also consume content eagerly all the time (leaving no option to user to make a choice).